internal/cmd/controller/controllers/clusterregistration/controller_test.go

```diff
@@ -0,0 +1,254 @@
+			Expect(newStatus.Granted).To(BeTrue())
+		})
+	})
+
+	Context("Cluster is missing", func() {
+		BeforeEach(func() {
+			request = &fleet.ClusterRegistration{
+				Spec: fleet.ClusterRegistrationSpec{
+					ClientID: "client-id",
+				},
+			}
+			status = fleet.ClusterRegistrationStatus{}
+
+			clusterCache.EXPECT().GetByIndex(gomock.Any(), gomock.Any()).AnyTimes().Return(nil, nil)
+			// NOTE(manno): code panics if cache.Get returns an error or nil
+			clusterCache.EXPECT().Get(gomock.Any(), gomock.Any()).AnyTimes().Return(nil, nil).Return(nil, notFound)
+			clusterClient.EXPECT().Get(gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes().Return(nil, nil)
+		})
+
+		When("cluster creation works", func() {
+			BeforeEach(func() {
+				clusterClient.EXPECT().Create(gomock.Any()).Return(nil, nil).Do(func(obj interface{}) {
+					switch cluster := obj.(type) {
+					case *fleet.Cluster:
+						Expect(cluster.Spec.ClientID).To(Equal("client-id"))
+					default:
+						Fail("unexpected type")
+					}
+				})
+			})
+
+			It("creates the missing cluster", func() {
+				objs, newStatus, err := h.OnChange(request, status)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(objs).To(BeEmpty())
+				Expect(newStatus.Granted).To(BeFalse())
+			})
+		})
+
+		When("cluster creation fails", func() {
+			BeforeEach(func() {
+				clusterClient.EXPECT().Create(gomock.Any()).Return(nil, anError)
+			})
+
+			It("returns an error", func() {
+				objs, newStatus, err := h.OnChange(request, status)
+				Expect(err).To(HaveOccurred())
+				Expect(objs).To(BeEmpty())
+				Expect(newStatus.Granted).To(BeFalse())
+			})
+		})
+	})
+
+	Context("Cluster exists", func() {
+		BeforeEach(func() {
+			request = &fleet.ClusterRegistration{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "request-1",
+					Namespace: "fleet-default",
+				},
+				Spec: fleet.ClusterRegistrationSpec{
+					ClientID: "client-id",
+				},
+			}
+
+			cluster = &fleet.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "cluster",
+					Namespace: "fleet-default",
+				},
+				Spec: fleet.ClusterSpec{
+					ClientID: "client-id",
+				},
+			}
+			status = fleet.ClusterRegistrationStatus{}
+
+			clusterCache.EXPECT().GetByIndex(gomock.Any(), gomock.Any()).AnyTimes().Return(nil, nil)
+			clusterCache.EXPECT().Get(gomock.Any(), gomock.Any()).AnyTimes().Return(nil, nil).Return(cluster, nil)
+		})
+
+		When("cluster status has no namespace", func() {
+			It("sets the cluster name into the registrations status", func() {
+				objs, newStatus, err := h.OnChange(request, status)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(objs).To(BeEmpty())
+				Expect(newStatus.Granted).To(BeFalse())
+				Expect(newStatus.ClusterName).To(Equal("cluster"))
+			})
+		})
+
+		When("service account does not exist", func() {
+			BeforeEach(func() {
+				cluster.Status = fleet.ClusterStatus{Namespace: "fleet-default"}
+				saCache.EXPECT().Get(gomock.Any(), gomock.Any()).AnyTimes().Return(nil, notFound)
```

Comment:
I'd be nice to assert the number of times methods should be called instead of always using `AnyTimes`